indigo/indigo/src/main/scala/indigo/platform/audio/AudioPlayer.scala

```diff
@@ -149,32 +151,36 @@ final class AudioPlayer(context: AudioContextProxy) {
   }
 
   private def updateSource(
-      sceneAudioSource: SceneAudioSource,
-      currentSource: AudioSourceState
-  ): Option[AudioSourceState] =
-    if (sceneAudioSource.bindingKey == currentSource.bindingKey) None
-    else
+      sceneAudioSource: Option[SceneAudioSource],
+      currentSource: Option[AudioSourceState]
+  ): Option[Option[AudioSourceState]] = (currentSource, sceneAudioSource) match {
+    case (None, None) =>
+      None
+    case (Some(playing), Some(next)) if playing.bindingKey == next.bindingKey =>
+      None
+    case (Some(playing), None) =>
+      playing.audioNodes.audioBufferSourceNode.stop()
+      Some(None)
+    case (_, Some(next)) =>
       Option {
-        sceneAudioSource.playbackPattern match {
-          case PlaybackPattern.Silent =>
-            currentSource.audioNodes.foreach(_.audioBufferSourceNode.stop(0))
-            new AudioSourceState(sceneAudioSource.bindingKey, None)
+        currentSource.foreach(_.audioNodes.audioBufferSourceNode.stop())
 
+        next.playbackPattern match {
           case PlaybackPattern.SingleTrackLoop(track) =>
-            currentSource.audioNodes.foreach(_.audioBufferSourceNode.stop(0))
-
             val nodes =
               findAudioDataByName(track.assetName)
-                .map(asset => setupNodes(asset, track.volume * sceneAudioSource.masterVolume, loop = true))
+                .map(asset => setupNodes(asset, track.volume * next.masterVolume, loop = true))
+                .get //throws if no asset found
```

Comment:
Commenting just for posterity: I think this is fine. I have this long running internal debate about what to do if an asset you need just isn't there, and bailing out is the only sensible thing to do at the moment. As a user you can check for the presence of your asset before you use it but it's a slightly awkward process. It's similar territory to having a backend Java service that's been running for weeks and then tries to reflectively load something that doesn't exist - what do you do?

Reply:
Also the `Outcome` type does let you handle stop-the-world issues like this (in theory).